# epiPALEOMIX Pipeline

A Fast, Accurate, and Automatic pipeline for generating nucleosome and methylation maps from high throughput sequencing data underlying ancient samples.


Recent developments in ancient DNA research have shown that not only ancient genomes but ancient epigenomes can be recovered from high-throughput DNA sequence data. Although direct methods, such as bisulfite DNA treatment [(Llamas et al. 2012)][llamas] and target enrichment for methylated epialleles (Seguin-Orlando et al. 2015), have been used, only indirect computational methods exploiting patterns of DNA degradation along the genome have succeeded in reconstructing genome-wide epigenetic maps (Pedersen et al. 2014, Gokhman et al. 2014). epiPALEOMIX implements the latter methods in a user-friendly and open-source pipeline, allowing non-experienced users to reconstruct both methylation and nucleosome maps from ancient DNA data. It requires BAM alignment files against reference genomes as input, which can easily be generated using the PALEOMIX pipeline, previously released by our group (Schubert et al. 2013). epiPALEOMIX has been developed by Kristian Hanghøj in the research group of Dr. Ludovic Orlando, at the Centre for GeoGenetics, Natural History Museum of Denmark, University of Copenhagen, Denmark.



For detailed instructions and an in-depth description of epiPALEOMIX, please refer to the companion Wiki site.

For questions, bug reports, and/or suggestions contact Kristian Hanghøj at k.hanghoej@snm.ku.dk

## Citation

Hanghøj K., ... , Orlando L. _Title_ 


## References

[llamas]: http://dx.doi.org/10.1371/journal.pone.0030226


***
[Go to wiki](https://bitbucket.org/khanghoj/epipaleomix/wiki/Home)
***
